File: winui4k-sample-gallery-in-java/src/main/java/com/appkitbox/winui4k/sample/gallery/StatusInfoPages.java
package com.appkitbox.winui4k.sample.gallery;

import com.appkitbox.winui4k.HorizontalAlignment;
import com.appkitbox.winui4k.InfoBarSeverity;
import com.appkitbox.winui4k.Orientation;
import com.appkitbox.winui4k.Symbol;
import com.appkitbox.winui4k.ToolTipPlacement;
import com.appkitbox.winui4k.VerticalAlignment;
import com.appkitbox.winui4k.WButton;
import com.appkitbox.winui4k.WComponent;
import com.appkitbox.winui4k.WInfoBadge;
import com.appkitbox.winui4k.WInfoBar;
import com.appkitbox.winui4k.WLabel;
import com.appkitbox.winui4k.WPanel;
import com.appkitbox.winui4k.WProgressBar;
import com.appkitbox.winui4k.WProgressRing;
import com.appkitbox.winui4k.WSlider;
import com.appkitbox.winui4k.WToolTip;

import kotlin.Unit;

/*
 * Status & info category: demo pages for InfoBadge / InfoBar / ProgressBar / ProgressRing / ToolTip.
 */
final class StatusInfoPages {
    private StatusInfoPages() {
    }

    // region InfoBadge

    /** The InfoBadge page: lines up demos for trying out WInfoBadge's various features. */
    static WComponent buildInfoBadgePage() {
        WPanel page = GalleryScaffold.buildPage(
                "InfoBadge",
                "A small badge that unobtrusively shows an unread count or draws attention. Try out WInfoBadge's various features.");

        page.add(buildInfoBadgeKindsExample());
        page.add(buildInfoBadgeDynamicExample());
        return page;
    }

    /** The kinds of badge: dot badge / numeric badge / icon badge. */
    private static WComponent buildInfoBadgeKindsExample() {
        WInfoBadge dot = new WInfoBadge(); // the default value of -1 makes it a dot badge

        WInfoBadge number = new WInfoBadge();
        number.setValue(5);

        WInfoBadge icon = new WInfoBadge();
        icon.setSymbolIcon(Symbol.MESSAGE);

        WPanel row = new WPanel(24.0, Orientation.HORIZONTAL);
        row.add(labeledColumn("Dot badge", dot));
        row.add(labeledColumn("Numeric badge", number));
        row.add(labeledColumn("Icon badge", icon));
        return GalleryScaffold.buildExample("Kinds of badge (Value = -1 / a number / IconSource)", row);
    }

    /** A dynamic badge: increment/reset the number with buttons. */
    private static WComponent buildInfoBadgeDynamicExample() {
        WInfoBadge badge = new WInfoBadge();
        badge.setValue(0);

        WButton addButton = new WButton("+1");
        addButton.addActionListener(() -> {
            badge.setValue(badge.getValue() + 1);
            return Unit.INSTANCE;
        });

        WButton clearButton = new WButton("Reset to 0");
        clearButton.addActionListener(() -> {
            badge.setValue(0);
            return Unit.INSTANCE;
        });

        WPanel row = new WPanel(16.0, Orientation.HORIZONTAL);
        row.add(labeledColumn("Unread", badge));
        addButton.setVerticalAlignment(VerticalAlignment.CENTER);
        row.add(addButton);
        clearButton.setVerticalAlignment(VerticalAlignment.CENTER);
        row.add(clearButton);
        return GalleryScaffold.buildExample("Updating the number (Value)", row);
    }

    /** A small column that stacks a control under a label (used by the badge demos). */
    private static WComponent labeledColumn(String label, WComponent control) {
        WPanel column = new WPanel(6.0, Orientation.VERTICAL);
        column.setHorizontalAlignment(HorizontalAlignment.LEFT);
        WLabel labelText = new WLabel(label);
        labelText.setForeground(GalleryTheme.TEXT_SECONDARY());
        column.add(labelText);
        control.setHorizontalAlignment(HorizontalAlignment.LEFT);
        column.add(control);
        return column;
    }

    // endregion

    // region InfoBar

    /** The InfoBar page: lines up demos for trying out WInfoBar's various features. */
    static WComponent buildInfoBarPage() {
        WPanel page = GalleryScaffold.buildPage(
                "InfoBar",
                "An inline notification bar that reports an in-app state change. Try out WInfoBar's various features.");

        page.add(buildInfoBarSeverityExample());
        page.add(buildInfoBarInteractiveExample());
        page.add(buildInfoBarActionExample());
        return page;
    }

    /** Severity: lines up all four of Informational / Success / Warning / Error. */
    private static WComponent buildInfoBarSeverityExample() {
        WPanel column = new WPanel(12.0, Orientation.VERTICAL);
        InfoBarSeverity[] severities = {
                InfoBarSeverity.INFORMATIONAL,
                InfoBarSeverity.SUCCESS,
                InfoBarSeverity.WARNING,
                InfoBarSeverity.ERROR,
        };
        String[][] texts = {
                {"Informational", "This is an informational message."},
                {"Success", "Your changes have been saved."},
                {"Warning", "The connection is unstable."},
                {"Error", "The save failed."},
        };
        for (int i = 0; i < severities.length; i++) {
            WInfoBar infoBar = new WInfoBar();
            infoBar.setSeverity(severities[i]);
            infoBar.setTitle(texts[i][0]);
            infoBar.setMessage(texts[i][1]);
            infoBar.setClosable(false);
            infoBar.setOpen(true);
            infoBar.setWidth(500.0);
            column.add(infoBar);
        }
        return GalleryScaffold.buildExample("Severity (Severity: Informational / Success / Warning / Error)", column);
    }

    /** Opening and closing: toggle isOpen with a button and receive close (x) button clicks. */
    private static WComponent buildInfoBarInteractiveExample() {
        WLabel result = new WLabel("The bar is closed");

        WInfoBar infoBar = new WInfoBar();
        infoBar.setSeverity(InfoBarSeverity.SUCCESS);
        infoBar.setTitle("Download complete");
        infoBar.setMessage("The file has finished downloading.");
        infoBar.setWidth(500.0);
        infoBar.addCloseButtonListener(() -> {
            result.setText("Closed via the close (x) button");
            return Unit.INSTANCE;
        });

        WButton toggleButton = new WButton("Open the bar");
        toggleButton.addActionListener(() -> {
            infoBar.setOpen(!infoBar.isOpen());
            toggleButton.setText(infoBar.isOpen() ? "Close the bar" : "Open the bar");
            result.setText(infoBar.isOpen() ? "The bar is open" : "The bar is closed");
            return Unit.INSTANCE;
        });

        WPanel column = new WPanel(12.0, Orientation.VERTICAL);
        WPanel row = new WPanel(16.0, Orientation.HORIZONTAL);
        row.add(toggleButton);
        result.setVerticalAlignment(VerticalAlignment.CENTER);
        row.add(result);
        column.add(row);
        column.add(infoBar);
        return GalleryScaffold.buildExample("Opening/closing and the close button (IsOpen / CloseButtonClick)", column);
    }

    /** An action button: places an action button next to the body text. */
    private static WComponent buildInfoBarActionExample() {
        WLabel result = new WLabel("");

        WButton actionButton = new WButton("Restart");
        actionButton.addActionListener(() -> {
            result.setText("A restart was requested");
            return Unit.INSTANCE;
        });

        WInfoBar infoBar = new WInfoBar();
        infoBar.setSeverity(InfoBarSeverity.WARNING);
        infoBar.setTitle("An update is pending");
        infoBar.setMessage("Restart the app to apply the changes.");
        infoBar.setActionButton(actionButton);
        infoBar.setClosable(false);
        infoBar.setOpen(true);
        infoBar.setWidth(500.0);

        WPanel column = new WPanel(12.0, Orientation.VERTICAL);
        column.add(infoBar);
        column.add(result);
        return GalleryScaffold.buildExample("Action button (ActionButton)", column);
    }

    // endregion

    // region ProgressBar

    /** The ProgressBar page: lines up demos for trying out WProgressBar's various features. */
    static WComponent buildProgressBarPage() {
        WPanel page = GalleryScaffold.buildPage(
                "ProgressBar",
                "A control that shows a task's progress as a bar. Try out WProgressBar's various features.");

        page.add(buildDeterminateProgressBarExample());
        page.add(buildIndeterminateProgressBarExample());
        page.add(buildProgressBarStateExample());
        return page;
    }

    /** Determinate progress: drag a slider to move value and reflect it on the bar. */
    private static WComponent buildDeterminateProgressBarExample() {
        WProgressBar progressBar = new WProgressBar(0.0, 100.0, 40.0);
        progressBar.setWidth(300.0);

        WLabel label = new WLabel("40 %");
        WSlider slider = new WSlider(0.0, 100.0, 40.0);
        slider.setWidth(300.0);
        slider.addChangeListener(value -> {
            progressBar.setValue(value);
            label.setText(value.intValue() + " %");
            return Unit.INSTANCE;
        });

        WPanel column = new WPanel(12.0, Orientation.VERTICAL);
        column.add(progressBar);
        column.add(slider);
        column.add(label);
        return GalleryScaffold.buildExample("Determinate progress (Value / Minimum / Maximum)", column);
    }

    /** Indeterminate progress: indicates a task whose completion time is unknown. */
    private static WComponent buildIndeterminateProgressBarExample() {
        WProgressBar progressBar = new WProgressBar(0.0, 100.0, 0.0);
        progressBar.setWidth(300.0);
        progressBar.setIndeterminate(true);

        WButton toggleButton = new WButton("Stop");
        toggleButton.addActionListener(() -> {
            progressBar.setIndeterminate(!progressBar.isIndeterminate());
            toggleButton.setText(progressBar.isIndeterminate() ? "Stop" : "Start");
            return Unit.INSTANCE;
        });

        WPanel column = new WPanel(12.0, Orientation.VERTICAL);
        column.add(progressBar);
        toggleButton.setHorizontalAlignment(HorizontalAlignment.LEFT);
        column.add(toggleButton);
        return GalleryScaffold.buildExample("Indeterminate progress (IsIndeterminate)", column);
    }

    /** Visual states: toggle the showError / showPaused visual states. */
    private static WComponent buildProgressBarStateExample() {
        WProgressBar progressBar = new WProgressBar(0.0, 100.0, 60.0);
        progressBar.setWidth(300.0);

        WButton pauseButton = new WButton("Pause");
        pauseButton.addActionListener(() -> {
            progressBar.setShowPaused(!progressBar.getShowPaused());
            pauseButton.setText(progressBar.getShowPaused() ? "Resume" : "Pause");
            return Unit.INSTANCE;
        });

        WButton errorButton = new WButton("Error");
        errorButton.addActionListener(() -> {
            progressBar.setShowError(!progressBar.getShowError());
            errorButton.setText(progressBar.getShowError() ? "Clear error" : "Error");
            return Unit.INSTANCE;
        });

        WPanel buttons = new WPanel(8.0, Orientation.HORIZONTAL);
        buttons.add(pauseButton);
        buttons.add(errorButton);

        WPanel column = new WPanel(12.0, Orientation.VERTICAL);
        column.add(progressBar);
        column.add(buttons);
        return GalleryScaffold.buildExample("Visual states (ShowPaused / ShowError)", column);
    }

    // endregion

    // region ProgressRing

    /** The ProgressRing page: lines up demos for trying out WProgressRing's various features. */
    static WComponent buildProgressRingPage() {
        WPanel page = GalleryScaffold.buildPage(
                "ProgressRing",
                "A control that shows an ongoing operation or progress as a circular ring. Try out WProgressRing's various features.");

        page.add(buildIndeterminateProgressRingExample());
        page.add(buildDeterminateProgressRingExample());
        return page;
    }

    /** Indeterminate progress: toggle the spinning animation with IsActive. */
    private static WComponent buildIndeterminateProgressRingExample() {
        WProgressRing progressRing = new WProgressRing();
        progressRing.setWidth(60.0);
        progressRing.setHeight(60.0);
        progressRing.setActive(true);

        WButton toggleButton = new WButton("Stop");
        toggleButton.addActionListener(() -> {
            progressRing.setActive(!progressRing.isActive());
            toggleButton.setText(progressRing.isActive() ? "Stop" : "Start");
            return Unit.INSTANCE;
        });

        WPanel row = new WPanel(16.0, Orientation.HORIZONTAL);
        row.add(progressRing);
        toggleButton.setVerticalAlignment(VerticalAlignment.CENTER);
        row.add(toggleButton);
        return GalleryScaffold.buildExample("Indeterminate progress (IsActive / IsIndeterminate)", row);
    }

    /** Determinate progress: drag a slider to move value and reflect it on the ring. */
    private static WComponent buildDeterminateProgressRingExample() {
        WProgressRing progressRing = new WProgressRing();
        progressRing.setWidth(60.0);
        progressRing.setHeight(60.0);
        progressRing.setIndeterminate(false);
        progressRing.setValue(40.0);

        WLabel label = new WLabel("40 %");
        WSlider slider = new WSlider(0.0, 100.0, 40.0);
        slider.setWidth(300.0);
        slider.addChangeListener(value -> {
            progressRing.setValue(value);
            label.setText(value.intValue() + " %");
            return Unit.INSTANCE;
        });

        WPanel row = new WPanel(16.0, Orientation.HORIZONTAL);
        row.add(progressRing);
        WPanel column = new WPanel(12.0, Orientation.VERTICAL);
        column.add(slider);
        column.add(label);
        column.setVerticalAlignment(VerticalAlignment.CENTER);
        row.add(column);
        return GalleryScaffold.buildExample("Determinate progress (IsIndeterminate = false / Value)", row);
    }

    // endregion

    // region ToolTip

    /** The ToolTip page: lines up demos for trying out WComponent.toolTip / WToolTip's various features. */
    static WComponent buildToolTipPage() {
        WPanel page = GalleryScaffold.buildPage(
                "ToolTip",
                "Shows a floating explanation of an element on hover. Try out the toolTip property and WToolTip.");

        page.add(buildSimpleToolTipExample());
        page.add(buildPlacementToolTipExample());
        page.add(buildRichToolTipExample());
        return page;
    }

    /** A string hint: hover over the toolTip property to see the explanation. */
    private static WComponent buildSimpleToolTipExample() {
        WButton button = new WButton("Hover over me");
        button.setToolTip("This is a plain string tooltip.");
        return GalleryScaffold.buildExample("A string hint (WComponent.toolTip)", button);
    }

    /** Specifying placement: choose WToolTip's placement from top / bottom / left / right. */
    private static WComponent buildPlacementToolTipExample() {
        WPanel row = new WPanel(12.0, Orientation.HORIZONTAL);
        String[] labels = {"Top", "Bottom", "Left", "Right"};
        ToolTipPlacement[] placements = {
                ToolTipPlacement.TOP,
                ToolTipPlacement.BOTTOM,
                ToolTipPlacement.LEFT,
                ToolTipPlacement.RIGHT,
        };
        for (int i = 0; i < labels.length; i++) {
            WButton button = new WButton(labels[i]);
            WToolTip toolTip = new WToolTip();
            toolTip.setText("Shown to the " + labels[i]);
            toolTip.setPlacement(placements[i]);
            button.setToolTip(toolTip);
            row.add(button);
        }
        return GalleryScaffold.buildExample("Specifying placement (WToolTip.placement)", row);
    }

    /** A non-string hint: put an arbitrary component in WToolTip. */
    private static WComponent buildRichToolTipExample() {
        WButton button = new WButton("Rich hint");

        WPanel content = new WPanel(4.0, Orientation.VERTICAL);
        WLabel heading = new WLabel("Heading");
        heading.setFontWeight(600);
        content.add(heading);
        WLabel description = new WLabel("The explanation can span multiple lines.");
        description.setForeground(GalleryTheme.TEXT_SECONDARY());
        content.add(description);

        WToolTip toolTip = new WToolTip();
        toolTip.setContent(content);
        button.setToolTip(toolTip);
        return GalleryScaffold.buildExample("A non-string hint (WToolTip.content)", button);
    }

    // endregion
}

File: winui4k-sample-gallery/src/main/kotlin/com/appkitbox/winui4k/sample/gallery/StatusInfoPages.kt
package com.appkitbox.winui4k.sample.gallery

import com.appkitbox.winui4k.HorizontalAlignment
import com.appkitbox.winui4k.InfoBarSeverity
import com.appkitbox.winui4k.Orientation
import com.appkitbox.winui4k.Symbol
import com.appkitbox.winui4k.ToolTipPlacement
import com.appkitbox.winui4k.VerticalAlignment
import com.appkitbox.winui4k.WButton
import com.appkitbox.winui4k.WComponent
import com.appkitbox.winui4k.WInfoBadge
import com.appkitbox.winui4k.WInfoBar
import com.appkitbox.winui4k.WLabel
import com.appkitbox.winui4k.WPanel
import com.appkitbox.winui4k.WProgressBar
import com.appkitbox.winui4k.WProgressRing
import com.appkitbox.winui4k.WSlider
import com.appkitbox.winui4k.WToolTip

/**
 * Status & info category: demo pages for InfoBadge / InfoBar / ProgressBar / ProgressRing / ToolTip.
 */

// region InfoBadge

/** The InfoBadge page: lines up demos for trying out WInfoBadge's various features. */
internal fun buildInfoBadgePage(): WComponent {
    val page = buildPage(
        "InfoBadge",
        "A small badge that unobtrusively shows an unread count or draws attention. Try out WInfoBadge's various features.",
    )

    page.add(buildInfoBadgeKindsExample())
    page.add(buildInfoBadgeDynamicExample())
    return page
}

/** The kinds of badge: dot badge / numeric badge / icon badge. */
private fun buildInfoBadgeKindsExample(): WComponent {
    val dot = WInfoBadge() // the default value = -1 is a dot badge

    val number = WInfoBadge()
    number.value = 5

    val icon = WInfoBadge()
    icon.setSymbolIcon(Symbol.MESSAGE)

    val row = WPanel(spacing = 24.0, orientation = Orientation.HORIZONTAL)
    row.add(labeledColumn("Dot badge", dot))
    row.add(labeledColumn("Numeric badge", number))
    row.add(labeledColumn("Icon badge", icon))
    return buildExample("Kinds of badge (Value = -1 / a number / IconSource)", row)
}

/** A dynamic badge: increment/reset the number with buttons. */
private fun buildInfoBadgeDynamicExample(): WComponent {
    val badge = WInfoBadge()
    badge.value = 0

    val addButton = WButton("+1")
    addButton.addActionListener { badge.value += 1 }

    val clearButton = WButton("Reset to 0")
    clearButton.addActionListener { badge.value = 0 }

    val row = WPanel(spacing = 16.0, orientation = Orientation.HORIZONTAL)
    row.add(labeledColumn("Unread", badge))
    row.add(addButton.also { it.verticalAlignment = VerticalAlignment.CENTER })
    row.add(clearButton.also { it.verticalAlignment = VerticalAlignment.CENTER })
    return buildExample("Updating the number (Value)", row)
}

/** A small column that stacks a control under a label (used by the badge demos). */
private fun labeledColumn(label: String, control: WComponent): WComponent {
    val column = WPanel(spacing = 6.0)
    column.horizontalAlignment = HorizontalAlignment.LEFT
    column.add(WLabel(label).also { it.foreground = TEXT_SECONDARY })
    column.add(control.also { it.horizontalAlignment = HorizontalAlignment.LEFT })
    return column
}

// endregion

// region InfoBar

/** The InfoBar page: lines up demos for trying out WInfoBar's various features. */
internal fun buildInfoBarPage(): WComponent {
    val page = buildPage(
        "InfoBar",
        "An inline notification bar that reports an in-app state change. Try out WInfoBar's various features.",
    )

    page.add(buildInfoBarSeverityExample())
    page.add(buildInfoBarInteractiveExample())
    page.add(buildInfoBarActionExample())
    return page
}

/** Severity: lines up all four of Informational / Success / Warning / Error. */
private fun buildInfoBarSeverityExample(): WComponent {
    val column = WPanel(spacing = 12.0)
    val severities = listOf(
        InfoBarSeverity.INFORMATIONAL to ("Informational" to "This is an informational message."),
        InfoBarSeverity.SUCCESS to ("Success" to "Your changes have been saved."),
        InfoBarSeverity.WARNING to ("Warning" to "The connection is unstable."),
        InfoBarSeverity.ERROR to ("Error" to "The save failed."),
    )
    for ((severity, text) in severities) {
        val infoBar = WInfoBar()
        infoBar.severity = severity
        infoBar.title = text.first
        infoBar.message = text.second
        infoBar.isClosable = false
        infoBar.isOpen = true
        infoBar.width = 500.0
        column.add(infoBar)
    }
    return buildExample("Severity (Severity: Informational / Success / Warning / Error)", column)
}

/** Opening and closing: toggle isOpen with a button and receive close (x) button clicks. */
private fun buildInfoBarInteractiveExample(): WComponent {
    val result = WLabel("The bar is closed")

    val infoBar = WInfoBar()
    infoBar.severity = InfoBarSeverity.SUCCESS
    infoBar.title = "Download complete"
    infoBar.message = "The file has finished downloading."
    infoBar.width = 500.0
    infoBar.addCloseButtonListener { result.text = "Closed via the close (x) button" }

    val toggleButton = WButton("Open the bar")
    toggleButton.addActionListener {
        infoBar.isOpen = !infoBar.isOpen
        toggleButton.text = if (infoBar.isOpen) "Close the bar" else "Open the bar"
        result.text = if (infoBar.isOpen) "The bar is open" else "The bar is closed"
    }

    val column = WPanel(spacing = 12.0)
    val row = WPanel(spacing = 16.0, orientation = Orientation.HORIZONTAL)
    row.add(toggleButton)
    row.add(result.also { it.verticalAlignment = VerticalAlignment.CENTER })
    column.add(row)
    column.add(infoBar)
    return buildExample("Opening/closing and the close button (IsOpen / CloseButtonClick)", column)
}

/** An action button: places an action button next to the body text. */
private fun buildInfoBarActionExample(): WComponent {
    val result = WLabel("")

    val actionButton = WButton("Restart")
    actionButton.addActionListener { result.text = "A restart was requested" }

    val infoBar = WInfoBar()
    infoBar.severity = InfoBarSeverity.WARNING
    infoBar.title = "An update is pending"
    infoBar.message = "Restart the app to apply the changes."
    infoBar.actionButton = actionButton
    infoBar.isClosable = false
    infoBar.isOpen = true
    infoBar.width = 500.0

    val column = WPanel(spacing = 12.0)
    column.add(infoBar)
    column.add(result)
    return buildExample("Action button (ActionButton)", column)
}

// endregion

// region ProgressBar

/** The ProgressBar page: lines up demos for trying out WProgressBar's various features. */
internal fun buildProgressBarPage(): WComponent {
    val page = buildPage(
        "ProgressBar",
        "A control that shows a task's progress as a bar. Try out WProgressBar's various features.",
    )

    page.add(buildDeterminateProgressBarExample())
    page.add(buildIndeterminateProgressBarExample())
    page.add(buildProgressBarStateExample())
    return page
}

/** Determinate progress: drag a slider to move value and reflect it on the bar. */
private fun buildDeterminateProgressBarExample(): WComponent {
    val progressBar = WProgressBar(value = 40.0)
    progressBar.width = 300.0

    val label = WLabel("40%")
    val slider = WSlider(value = 40.0)
    slider.width = 300.0
    slider.addChangeListener { value ->
        progressBar.value = value
        label.text = "${value.toInt()}%"
    }

    val column = WPanel(spacing = 12.0)
    column.add(progressBar)
    column.add(slider)
    column.add(label)
    return buildExample("Determinate progress (Value / Minimum / Maximum)", column)
}

/** Indeterminate progress: indicates a task whose completion time is unknown. */
private fun buildIndeterminateProgressBarExample(): WComponent {
    val progressBar = WProgressBar()
    progressBar.width = 300.0
    progressBar.isIndeterminate = true

    val toggleButton = WButton("Stop")
    toggleButton.addActionListener {
        progressBar.isIndeterminate = !progressBar.isIndeterminate
        toggleButton.text = if (progressBar.isIndeterminate) "Stop" else "Start"
    }

    val column = WPanel(spacing = 12.0)
    column.add(progressBar)
    column.add(toggleButton.also { it.horizontalAlignment = HorizontalAlignment.LEFT })
    return buildExample("Indeterminate progress (IsIndeterminate)", column)
}

/** Visual states: toggle the showError / showPaused visual states. */
private fun buildProgressBarStateExample(): WComponent {
    val progressBar = WProgressBar(value = 60.0)
    progressBar.width = 300.0

    val pauseButton = WButton("Pause")
    pauseButton.addActionListener {
        progressBar.showPaused = !progressBar.showPaused
        pauseButton.text = if (progressBar.showPaused) "Resume" else "Pause"
    }

    val errorButton = WButton("Error")
    errorButton.addActionListener {
        progressBar.showError = !progressBar.showError
        errorButton.text = if (progressBar.showError) "Clear error" else "Error"
    }

    val buttons = WPanel(spacing = 8.0, orientation = Orientation.HORIZONTAL)
    buttons.add(pauseButton)
    buttons.add(errorButton)

    val column = WPanel(spacing = 12.0)
    column.add(progressBar)
    column.add(buttons)
    return buildExample("Visual states (ShowPaused / ShowError)", column)
}

// endregion

// region ProgressRing

/** The ProgressRing page: lines up demos for trying out WProgressRing's various features. */
internal fun buildProgressRingPage(): WComponent {
    val page = buildPage(
        "ProgressRing",
        "A control that shows an ongoing operation or progress as a circular ring. Try out WProgressRing's various features.",
    )

    page.add(buildIndeterminateProgressRingExample())
    page.add(buildDeterminateProgressRingExample())
    return page
}

/** Indeterminate progress: toggle the spinning animation with IsActive. */
private fun buildIndeterminateProgressRingExample(): WComponent {
    val progressRing = WProgressRing()
    progressRing.width = 60.0
    progressRing.height = 60.0
    progressRing.isActive = true

    val toggleButton = WButton("Stop")
    toggleButton.addActionListener {
        progressRing.isActive = !progressRing.isActive
        toggleButton.text = if (progressRing.isActive) "Stop" else "Start"
    }

    val row = WPanel(spacing = 16.0, orientation = Orientation.HORIZONTAL)
    row.add(progressRing)
    row.add(toggleButton.also { it.verticalAlignment = VerticalAlignment.CENTER })
    return buildExample("Indeterminate progress (IsActive / IsIndeterminate)", row)
}

/** Determinate progress: drag a slider to move value and reflect it on the ring. */
private fun buildDeterminateProgressRingExample(): WComponent {
    val progressRing = WProgressRing()
    progressRing.width = 60.0
    progressRing.height = 60.0
    progressRing.isIndeterminate = false
    progressRing.value = 40.0

    val label = WLabel("40%")
    val slider = WSlider(value = 40.0)
    slider.width = 300.0
    slider.addChangeListener { value ->
        progressRing.value = value
        label.text = "${value.toInt()}%"
    }

    val row = WPanel(spacing = 16.0, orientation = Orientation.HORIZONTAL)
    row.add(progressRing)
    val column = WPanel(spacing = 12.0)
    column.add(slider)
    column.add(label)
    row.add(column.also { it.verticalAlignment = VerticalAlignment.CENTER })
    return buildExample("Determinate progress (IsIndeterminate = false / Value)", row)
}

// endregion

// region ToolTip

/** The ToolTip page: lines up demos for trying out WComponent.toolTip / WToolTip's various features. */
internal fun buildToolTipPage(): WComponent {
    val page = buildPage(
        "ToolTip",
        "Shows a floating explanation of an element on hover. Try out the toolTip property and WToolTip.",
    )

    page.add(buildSimpleToolTipExample())
    page.add(buildPlacementToolTipExample())
    page.add(buildRichToolTipExample())
    return page
}

/** A string hint: hover over the toolTip property to see the explanation. */
private fun buildSimpleToolTipExample(): WComponent {
    val button = WButton("Hover over me")
    button.toolTip = "This is a plain string tooltip."
    return buildExample("A string hint (WComponent.toolTip)", button)
}

/** Specifying placement: choose WToolTip's placement from top / bottom / left / right. */
private fun buildPlacementToolTipExample(): WComponent {
    val row = WPanel(spacing = 12.0, orientation = Orientation.HORIZONTAL)
    val placements = listOf(
        "Top" to ToolTipPlacement.TOP,
        "Bottom" to ToolTipPlacement.BOTTOM,
        "Left" to ToolTipPlacement.LEFT,
        "Right" to ToolTipPlacement.RIGHT,
    )
    for ((label, placement) in placements) {
        val button = WButton(label)
        val toolTip = WToolTip()
        toolTip.text = "Shown to the $label"
        toolTip.placement = placement
        button.setToolTip(toolTip)
        row.add(button)
    }
    return buildExample("Specifying placement (WToolTip.placement)", row)
}

/** A non-string hint: put an arbitrary component in WToolTip. */
private fun buildRichToolTipExample(): WComponent {
    val button = WButton("Rich hint")

    val content = WPanel(spacing = 4.0)
    content.add(WLabel("Heading").also { it.fontWeight = 600 })
    content.add(WLabel("The explanation can span multiple lines.").also { it.foreground = TEXT_SECONDARY })

    val toolTip = WToolTip()
    toolTip.content = content
    button.setToolTip(toolTip)
    return buildExample("A non-string hint (WToolTip.content)", button)
}

// endregion
